tesseract/evm/src/gas_oracle.rs

```diff
@@ -397,12 +455,63 @@ mod test {
 		get_l2_data_cost, parse_to_27_decimals, ARBITRUM_SEPOLIA_CHAIN_ID, BSC_TESTNET_CHAIN_ID,
 		GNOSIS_CHAIN_ID, OPTIMISM_SEPOLIA_CHAIN_ID, POLYGON_TESTNET_CHAIN_ID, SEPOLIA_CHAIN_ID,
 	};
+	use crate::gas_oracle::GasBreakdown; 
 	use ethers::{prelude::Provider, providers::Http, utils::parse_units};
 	use ismp::host::StateMachine;
 	use primitive_types::U256;
 	use std::sync::Arc;
 	use tesseract_primitives::Cost;
 
+	// Add to existing test module
+	async fn setup_test_provider() -> Arc<Provider<Http>> {
+		// For now we'll use a real provider just to test our new logic
+		let ethereum_rpc_uri = std::env::var("OP_URL").expect("op url is not set in .env");
+		let provider = Provider::<Http>::try_from(ethereum_rpc_uri).unwrap();
+		Arc::new(provider)
+	}
+	// Helper function to create mock data
+    fn setup_test_data() -> (StateMachine, Arc<Provider<Http>>, String) {
+        let provider = Provider::<Http>::try_from("http://mock.url").unwrap();
+        
+        (
+            StateMachine::Evm(OPTIMISM_SEPOLIA_CHAIN_ID),
+            Arc::new(provider),
+            "mock_api_key".to_string()
+        )
+    }
+
+    #[tokio::test]
+    async fn test_optimism_total_gas_calculation() {
```

Comment:
It would be much better to test this using OpSepolia directly

Reply:
Hi @Wizdave97,

I've added the requested comments and references in the code, including links to the OpStack repository for additional context. Thanks again for the feedback—it's much appreciated!

Regarding testing against OpSepolia, I've currently limited the tests to unit-level logic simulations. Due to time constraints and other pending PRs I need to address, I haven't yet integrated direct end-to-end testing on OpSepolia. However, I definitely see value in adding those tests and will plan to revisit this when I have more bandwidth. It would be great to ensure full coverage and validate the changes in a live environment as well.

Reply:
Hi @Wizdave97, I've now integrated direct testing against OpSepolia as requested. The tests run successfully against the OpSepolia endpoint, verifying both the L2 data cost calculations and the gas estimation logic in a live environment. Additionally, I've refined the code comments and references to the OpStack/Ecotone specifications for clarity. Let me know if there's anything else you'd like addressed! 
<img width="611" alt="Screenshot 2024-12-17 at 5 31 53 PM" src="https://github.com/user-attachments/assets/ba2d8ec8-eeba-4547-8151-ca4f708ccbb8" />

Reply:
Hi @Wizdave97, I've now integrated direct testing against OpSepolia as requested. The tests run successfully against the OpSepolia endpoint, verifying both the L2 data cost calculations and the gas estimation logic in a live environment. Additionally, I've refined the code comments and references to the OpStack/Ecotone specifications for clarity. Let me know if there's anything else you'd like addressed! 
<img width="611" alt="Screenshot 2024-12-17 at 5 31 53 PM" src="https://github.com/user-attachments/assets/ba2d8ec8-eeba-4547-8151-ca4f708ccb